pkg/dns/vips/tag_first_virtual_outbound_view.go

```diff
@@ -0,0 +1,220 @@
+// Now we don't duplicate tags, especially kuma.io/service tag.
+// This model produces configuration which is around 50% characters smaller than default one
+type TagFirstVirtualOutboundView struct {
+	PerService map[string]*TagFirstOutbound `json:"v,omitempty"`
+}
+
+type TagFirstOutbound struct {
+	Outbounds []*Outbound `json:"oe,omitempty"`
+}
+
+type Outbound struct {
+	Tags         map[string]string `json:"t,omitempty"`
+	AddressPorts []AddressPort     `json:"ap,omitempty"`
+}
+
+type AddressPort struct {
+	Address *string `json:"a,omitempty"`
+	Domain  *string `json:"d,omitempty"`
+	Port    *uint32 `json:"p,omitempty"`
+	Origin  *string `json:"o,omitempty"`
+}
+
+func (t *TagFirstOutbound) lookup(tags map[string]string) (*Outbound, bool) {
+	for _, i := range t.Outbounds {
+		if reflect.DeepEqual(i.Tags, tags) {
+			return i, true
+		}
+	}
+	return nil, false
+}
+
+func NewEmptyTagFirstOutboundView() *TagFirstVirtualOutboundView {
+	return &TagFirstVirtualOutboundView{
+		PerService: map[string]*TagFirstOutbound{},
+	}
+}
+
+func NewTagFirstOutboundView(virtualOutboundMeshView *VirtualOutboundMeshView) *TagFirstVirtualOutboundView {
+	if virtualOutboundMeshView == nil {
+		return nil
+	}
+
+	view := NewEmptyTagFirstOutboundView()
+
+	for he, vo := range virtualOutboundMeshView.byHostname {
+		var domain *string
+		if he.Type == Host || he.Type == FullyQualifiedDomain {
+			if govalidator.IsDNSName(he.Name) {
+				domain = pointer.To(he.Name)
+			}
+		}
+		for _, oe := range vo.Outbounds {
+			svc := oe.TagSet[mesh_proto.ServiceTag]
+			port := getOrNil(oe.Port)
+			origin := minifyOrigin(oe.Origin)
+			tagFirst, ok := view.PerService[svc]
+			if !ok {
+				outbound := &Outbound{
+					Tags: copyTagsWithoutSvc(oe.TagSet),
+					AddressPorts: []AddressPort{{
+						Address: &vo.Address,
+						Domain:  domain,
+						Port:    port,
+						Origin:  pointer.To(origin),
+					}},
+				}
+				view.PerService[svc] = &TagFirstOutbound{Outbounds: []*Outbound{outbound}}
+			} else {
+				cTags := copyTagsWithoutSvc(oe.TagSet)
+				outbound, ok := tagFirst.lookup(cTags)
+				if ok {
+					outbound.AddressPorts = append(outbound.AddressPorts, AddressPort{
+						Address: &vo.Address,
+						Domain:  domain,
+						Port:    port,
+						Origin:  pointer.To(origin),
+					})
+				} else {
+					tagFirst.Outbounds = append(tagFirst.Outbounds, &Outbound{
+						Tags: cTags,
+						AddressPorts: []AddressPort{{
+							Address: &vo.Address,
+							Domain:  domain,
+							Port:    port,
+							Origin:  pointer.To(origin),
+						}},
+					})
+				}
+			}
+		}
+	}
+
+	return view
+}
+
+func (t *TagFirstVirtualOutboundView) ToVirtualOutboundView() *VirtualOutboundMeshView {
+	view := NewEmptyVirtualOutboundView()
+
+	for svc, tagFirstOutbound := range t.PerService {
+		for _, outbound := range tagFirstOutbound.Outbounds {
+			for _, ap := range outbound.AddressPorts {
+				entryType := originToEntryType(pointer.Deref(ap.Origin))
```

Comment:
is this O(l*m\*n) conversion potentially a performance concern? IIUC the existing VIP view is kind of the "default" and the new tag first is converted to it on read from the config map? Should/can it be the other way around?

Reply:
I think this is not go to strategy, this is mostly to evaluate if this fix our issues with size of the configMap on k8s, and in the future there should be no conversion. We should be using of the model across the whole project without mixing them